File: backend/src/shared/protocol/misc.cc
// Copyright 2023 The Z2K Plus+ Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "z2kplus/backend/shared/protocol/misc.h"

#include <iostream>
#include "kosak/coding/coding.h"
#include "kosak/coding/myjson.h"

using kosak::coding::streamf;

namespace z2kplus::backend::shared::protocol {
std::ostream &operator<<(std::ostream &s, const Estimate &o) {
  return streamf(s, "[%o, %o]", o.count_, o.exact_);
}

DEFINE_TYPICAL_JSON(Estimate, count_, exact_);

Estimates Estimates::create(size_t frontCount, size_t backCount, bool frontIsExact, bool backIsExact) {
  Estimate front(frontCount, frontIsExact);
  Estimate back(backCount, backIsExact);
  return {front, back};
}

Estimates::Estimates(const Estimate &front, const Estimate &back) :
    front_(front), back_(back) {}

bool operator==(const Estimates &lhs, const Estimates &rhs) {
  return lhs.front_ == rhs.front_ && lhs.back_ == rhs.back_;
}

std::ostream &operator<<(std::ostream &s, const Estimates &o) {
  return streamf(s, "[front=%o, back=%o]", o.front_, o.back_);
}

DEFINE_TYPICAL_JSON(Estimates, front_, back_);

Filter::Filter() = default;
Filter::Filter(
    std::optional<std::string> sender, std::optional<std::string> instanceExact,
    std::optional<std::string> instancePrefix, bool strong) :
    sender_(std::move(sender)), instanceExact_(std::move(instanceExact)),
    instancePrefix_(std::move(instancePrefix)), strong_(strong) {}
Filter::Filter(const Filter &other) = default;
Filter::Filter(Filter &&other) noexcept = default;
Filter &Filter::operator=(const Filter &other) = default;
Filter &Filter::operator=(Filter &&other) noexcept = default;
Filter::~Filter() = default;

std::ostream &operator<<(std::ostream &s, const Filter &o) {
  return streamf(s, "Filter(%o,%o,%o,%o)", o.sender_, o.instanceExact_, o.instancePrefix_, o.strong_);
}

DEFINE_TYPICAL_JSON(Filter, sender_, instanceExact_, instancePrefix_, strong_);
}  // namespace z2kplus::backend::shared::protocol

File: frontend/shared/protocol/misc.ts
// Copyright 2023-2024 The Z2K Plus+ Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  assertAndDestructure1, assertAndDestructure2, assertAndDestructure5, assertArray, assertArrayOfLength,
  assertBoolean, assertNumber, assertString, optionalToJson, optionalTryParseJson,
} from "../json_util";
import {booleanCompare, IComparable, intercalate} from "../utility";

export class Unit implements IComparable<Unit> {
  private static readonly instance = new Unit();

  static create() {
    return Unit.instance;
  }

  private constructor() {}

  toJson() {
    return [];
  }

  compareTo(other: Unit): number {
    return 0;
  }

  static tryParseJson(item: any) {
    assertArrayOfLength(item, 0);
    return Unit.create();
  }

  toString() {
    return `Unit()`;
  }
}

export class Estimate {
  constructor(readonly count: number, readonly exact: boolean) {
  }

  toJson() {
    return [this.count, this.exact];
  }

  static tryParseJson(item: any) {
    let [c, lb] = assertAndDestructure2(item, assertNumber, assertBoolean);
    return new Estimate(c, lb);
  }

  compareTo(other: Estimate) {
    let difference = this.count - other.count;
    if (difference !== 0) return difference;
    difference = booleanCompare(this.exact, other.exact);
    return difference;
  }

  toString() {
    return `Estimate(${this.count}, ${this.exact})`;
  }
}

export class Estimates {
  constructor(readonly front: Estimate, readonly back: Estimate) {
  }

  toJson() {
    return [this.front.toJson(), this.back.toJson()];
  }

  static tryParseJson(item: any) {
    let [f, b] = assertAndDestructure2(item, Estimate.tryParseJson, Estimate.tryParseJson);
    return new Estimates(f, b);
  }

  compareTo(other: Estimates) {
    let difference: number;
    difference = this.front.compareTo(other.front);
    if (difference !== 0) return difference;
    difference = this.back.compareTo(other.back);
    return difference;
  }

  toString() {
    return `Estimate(${this.front}, ${this.back})`;
  }
}

export class Filter {
  constructor(
      readonly sender: string | undefined, readonly instanceExact: string | undefined,
      readonly instancePrefix: string | undefined, readonly strong: boolean,
      readonly expirationSecs: number) {
  }

  toJson() {
    return [
        optionalToJson(this.sender, i => i),
        optionalToJson(this.instanceExact, i => i),
        optionalToJson(this.instancePrefix, i => i),
        this.strong,
        this.expirationSecs];
  }

  static tryParseJson(item: any) {
    let [sd, ix, ip, st, ex] = assertAndDestructure5(item,
        i => optionalTryParseJson(i, assertString),
        i => optionalTryParseJson(i, assertString),
        i => optionalTryParseJson(i, assertString),
        assertBoolean,
        assertNumber);
    return new Filter(sd, ix, ip, st, ex);
  }

  toString() {
    return `Filter(${this.sender}, ${this.instanceExact}, ${this.instancePrefix}, ${this.strong}, ` +
      `${this.expirationSecs})`;
  }
}
//
// export class Filters {
//   constructor(readonly filters: Filter[]) {}
//
//   toJson() {
//     return [this.filters.map(f => f.toJson())];
//   }
//
//   static tryParseJson(item: any) {
//     const [fs] = assertAndDestructure1(item, assertArray);
//     const filters = fs.map(Filter.tryParseJson);
//     return new Filters(filters);
//   }
//
//   toString() {
//     const text = intercalate(", ", x => x.toString(), this.filters);
//     return `Filters(${text})`;
//   }
// }
